src/System.Binary.Base64/System/Binary/Base64Decoder.cs

```diff
@@ -0,0 +1,266 @@
+            Unsafe.Add(ref destBytes, 2) = (byte)i0;
+        }
+
+        public TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+        {
+            ref byte srcBytes = ref source.DangerousGetPinnableReference();
+            ref byte destBytes = ref destination.DangerousGetPinnableReference();
+
+            int srcLength = source.Length / 4 * 4;  // only decode input up to the closest multiple of 4.
+            int destLength = destination.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+
+            if (source.Length == 0) goto DoneExit;
+
+            ref sbyte decodingMap = ref s_decodingMap[0];
+
+            while (sourceIndex < srcLength - 4) // TODO: you can make this more efficient by using unsafe pointer as the loop limit
+            {
+                int i0 = Decode(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (i0 < 0) goto InvalidExit;
+                if (destIndex > destLength - 3) goto DestinationSmallExit; // TODO: you can get rid off this line by including it in loop limit
+                WriteThreeBytes(ref Unsafe.Add(ref destBytes, destIndex), i0);
+                destIndex += 3;
+                sourceIndex += 4;
+            }
+
+            if (sourceIndex >= srcLength) goto NeedMoreExit;
+
+            int padding = 0;
+
+            if (Unsafe.Add(ref srcBytes, srcLength - 1) == s_encodingPad)
+            {
+                padding = (Unsafe.Add(ref srcBytes, srcLength - 2) == s_encodingPad) ? 2 : 1;
+            }
+
+            if (padding == 1)
+            {
+                int i0 = DecodePadByOne(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (i0 < 0) goto InvalidExit;
+                if (destIndex > destLength - 2) goto DestinationSmallExit;
+                Unsafe.Add(ref destBytes, destIndex) = (byte)(i0 >> 16);
+                Unsafe.Add(ref destBytes, destIndex + 1) = (byte)(i0 >> 8);
+                destIndex += 2;
+                sourceIndex += 4;
+            }
+            else if (padding == 2)
+            {
+                int i0 = DecodePadByTwo(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (i0 < 0) goto InvalidExit;
+                if (destIndex > destLength - 1) goto DestinationSmallExit;
+                Unsafe.Add(ref destBytes, destIndex) = (byte)(i0 >> 16);
+                destIndex += 1;
+                sourceIndex += 4;
+            }
+            else
+            {
+                int i0 = Decode(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (i0 < 0) goto InvalidExit;
+                if (destIndex > destLength - 3) goto DestinationSmallExit;
+                WriteThreeBytes(ref Unsafe.Add(ref destBytes, destIndex), i0);
+                destIndex += 3;
+                sourceIndex += 4;
```

Comment:
`sourceIndex += 4;` is same in every if case, so you can move it after the if chain to save a bit code size.